### PR TITLE
Crafting/trade goods filter update

### PR DESCRIPTION
**Pull request by Cidan**
This CL adds the ability to further filter trade goods automatic categories with an optional breakdown by expansion pack. This also changes the name of the trade good categories so they reflect a profession instead of the type of item it is, i.e. "Mining" instead of "Stone and Metal", "Tailoring" instead of "Cloth".

Name normalization was decided due to the mismatched naming between trade good functions in the Blizzard API.

### CODE REVIEW

**Comment by Cidan** [comment]
I also want to add a quick toggle/quad-state toggle to the top of the reagent bank window to switch between the four possible combinations of tradeskill + xpac splitting, before this goes live.

**Comment by Talyrius** [comment]
>This also changes the name of the trade good categories so they reflect a profession instead of the type of item it is, i.e. "Mining" instead of "Stone and Metal", "Tailoring" instead of "Cloth".
>
> Name normalization was decided due to the mismatched naming between trade good functions in the Blizzard API.

😬

I think it makes much more sense that trade goods continue to be categorized by their type rather than the profession they're primarily used for or gathered by as they aren't strictly profession specific.

> I also want to add a quick toggle/quad-state toggle to the top of the reagent bank window to switch between the four possible combinations of tradeskill + xpac splitting, before this goes live.

Yes, this would come in handy for temporarily toggling it on and off if you've forgotten which expansion something has come from, but you don't always want your trade goods split up or your bag filled with the names of various expansions. Regarding that, maybe they can be optionally abbreviated or perhaps even better yet—replaced with textures, replete with alt text: 🤔
World of Warcraft, The Burning Crusade, Wrath of the Lich King, Cataclysm, Mists of Pandaria, Warlords of Draenor, Legion, Battle for Azeroth, Shadowlands, Dragonflight

**Comment by Cidan** [comment]
Yeah, I went back and forth on that one a bit. Recently, Blizzard added a professions Enum and item type that explicitly lists if an item is used in a specific profession. This is used for things like mining picks, rods, etc. I was hoping to align with that Enum, but perhaps you're right here. I'll go ahead and use the trade goods name.

I think the toggle will come at a later date.

re: section names, this is something I've been thinking about for a long time now. I want to figure out a more generalized solution for section names that are very long, but only have one or two items, This one will be out of scope for now. Let me make the change to flip trade skill names back.

**Comment by Talyrius** [comment]
Also, we should try and use Blizzard's localized returns/constants whenever possible to avoid having to localize them ourselves.

**Comment by Cidan** [comment]
I reverted the tradeskill name change, and I'm assuming your comment around localized returns was based on the expansion names. I couldn't find constants with the expansion names in it, just enum's. Do you happen to know where they live?

**Comment by Cidan** [comment]
p.s. unless you have any other comments, I'm g2g with this PR.

**Comment by Talyrius** [comment]
> Do you happen to know where they live?

No, but I can try and find them. Localization will definitely be asked for.

> p.s. unless you have any other comments, I'm g2g with this PR.

I haven't had a chance to look at this in-game yet—can do so later tonight (also PST).

**Comment by Cidan** [comment]
Alright, cool. I looked around Gethe's UI dump, and nothing turned up. Kind of surprising, actually.

**Comment by Cidan** [comment]
Meo pointed me in the right direction for the expansion name localized constants. Updated the map with them.

**Comment by Talyrius** [comment]
Nice. 👍

**Comment by Cidan** [comment]
@Talyrius Did you get a chance to check this out? I'm eager to merge this so I can continue the annotation work in the other branch.

File: core/ItemDatabase.lua [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+--]]
+
+local addonName, addon = ...
+
+addon.ItemDatabase = addon:NewModule('ItemDatabase', 'AceEvent-3.0')
+local db = addon.ItemDatabase
+
+function db:ReagentData(slotData)
+  if not slotData.isCraftingReagent then return false end
+  if not slotData.classID == LE_ITEM_CLASS_TRADEGOODS then return false end
```
Comment (Talyrius) [inline core/ItemDatabase.lua:29]: `LE_ITEM_CLASS_TRADEGOODS` has been deprecated.

**Review by Talyrius — CHANGES_REQUESTED** [review]
Sorry, I never have as much free time as I aspire to have! 😢

If you don't mind, I'm going to push a commit that addresses a few things.

File: core/DefaultFilters.lua [hunk 2]
```diff
@@ -363,13 +371,15 @@ function addon:SetupDefaultFilters()
...
+
+			local reagentData = addon.ItemDatabase:ReagentData(slotData)
+
 			if self.db.profile.splitBySubclass[class] then
-				return subclass, class
+				return (self.db.profile.splitExpansion and reagentData and reagentData.expansionName .. " - " ..subclass) or (reagentData and reagentData.profession) or subclass, class
 			else
-				return class
+				return (self.db.profile.splitExpansion and reagentData and reagentData.expansionName .. " - " ..class) or class
 			end
```
[the remark this answers is not in the document]
Comment (Cidan) [reply core/DefaultFilters.lua]: Resolving this since it has changed

File: core/ItemDatabase.lua (hunk 1 above)
Comment (Cidan) [reply core/ItemDatabase.lua]: Resolving this since this has changed.

File: core/Constants.lua [hunk 3]
```diff
@@ -76,6 +76,45 @@ for _, bags in ipairs { BAGS, BANK } do
...
+	[11] = "Other",
+	[12] = CHARACTER_PROFESSION_ENCHANTING,
+	[13] = "Materials",
+	[14] = "Item Enchantement",
+	[15] = WEAPON_ENCHANTMENT,
+	[16] = CHARACTER_PROFESSION_INSCRIPTION,
+	[17] = "Explosives and Devices",
+	[18] = OPTIONAL_REAGENT_TUTORIAL_TOOLTIP_TITLE,
+	[19] = "Finishing Reagents"
+}
```
[the remark this answers is not in the document]
Comment (Cidan) [reply core/Constants.lua]: Resolving this, your solution is much better!

File: core/DefaultFilters.lua [hunk 4]
```diff
@@ -372,12 +366,14 @@ function addon:SetupDefaultFilters()
...
+			--TODO: Implement an option to override `subclass` with professions from our `addon.TRADESKILL_MAP`?
 			local reagentData = addon.ItemDatabase:ReagentData(slotData)
-
 			if self.db.profile.splitBySubclass[class] then
-				return (self.db.profile.splitExpansion and reagentData and reagentData.expansionName .. " - " ..subclass) or (reagentData and reagentData.profession) or subclass, class
+				return (self.db.profile.splitExpansion and reagentData and subclass.." ("..reagentData.expacName..")")
+				or subclass, class
 			else
-				return (self.db.profile.splitExpansion and reagentData and reagentData.expansionName .. " - " ..class) or class
+				return (self.db.profile.splitExpansion and reagentData and class.." ("..reagentData.expacName..")")
```
Comment (Cidan) [inline core/DefaultFilters.lua:375]: nit: I don't like using `(text)` in the English language, as it subordinates one clause to another. I think either the previous `-` is more fitting, or perhaps `class: expansion`.

**Comment by Cidan** [comment]
Your changes are great, and I learned quite a bit from your changes on this. Thank you!

**Comment by Cidan** [comment]
Unrelated, I never understood this about WoW, but why is `_G.` explicitly called when using globals, when globals are... well, global?

File: core/DefaultFilters.lua (hunk 4 above)
Comment (Talyrius) [reply core/DefaultFilters.lua:375]: Yes, either is fine by me. My preference would be for `class: expansion` or even `class, expansion`.

**Comment by Cidan** [comment]
I just got an answer on `_G`, see my other PR for comments on this.

**Comment by Talyrius** [comment]
Thanks!

Using `_G.` isn't strictly necessary, but if you use something like Luacheck, it will pester you about making the clarification.

**Comment by Cidan** [comment]
Alright, I changed it to `:` and tested it in game, it looks great. When you're ready, approve and I'll merge + release. :)

**Review by Talyrius — APPROVED** [review]

**Comment by JourneyOver** [comment]
Not sure if this is an issue on just my side or if this is an issue with adibags but upon updating to 1.10.3 I started getting this lua error:

```
Interface/AddOns/AdiBags/core/DefaultFilters.lua:370: attempt to index field 'ItemDatabase' (a nil value)
Count: 99

Call Stack:
[string "@Interface/AddOns/AdiBags/core/Utility.lua"]:90: in function <Interface/AddOns/AdiBags/core/Utility.lua:86>
[string "=(tail call)"]: ?
[string "@Interface/AddOns/AdiBags/core/Filters.lua"]:146: in function `Filter'
[string "@Interface/AddOns/AdiBags/widgets/ContainerFrame.lua"]:844: in function `FilterSlot'
[string "@Interface/AddOns/AdiBags/widgets/ContainerFrame.lua"]:878: in function `DispatchItem'
[string "@Interface/AddOns/AdiBags/widgets/ContainerFrame.lua"]:1047: in function `RedispatchAllItems'
[string "@Interface/AddOns/AdiBags/widgets/ContainerFrame.lua"]:1205: in function `FullUpdate'
[string "@Interface/AddOns/AdiBags/widgets/ContainerFrame.lua"]:640: in function <Interface/AddOns/AdiBags/widgets/ContainerFrame.lua:626>
[string "=[C]"]: ?
[string "@Interface/AddOns/AdiBags/core/Utility.lua"]:102: in function <Interface/AddOns/AdiBags/core/Utility.lua:94>
[string "@Interface/AddOns/AdiBags/widgets/LayeredRegion.lua"]:119: in function <Interface/AddOns/AdiBags/widgets/LayeredRegion.lua:108>
```

**Comment by Cidan** [comment]
Er. What version of WoW?

**Comment by Cidan** [comment]
I just yanked the update from curseforge in an abundance of caution -- WoW authentication servers are down and I can't login, so I can't test, but this was working for both @Talyrius and I.

**Comment by SpareSimian** [comment]
@JourneyOver , that one's a side effect of issue #904 and a fix is just waiting on the auth servers t to come back up for testing.

**Comment by JourneyOver** [comment]
@Cidan I'm on retail.

@SpareSimian Good to know, I'll stay downgraded to 1.10.2 until the fix has been pushed. Crazy how blizz has been having so much issues since yesterday xD

**Comment by Cidan** [comment]
@JourneyOver Fixed has been pushed, you should be able to roll forward to 1.10.4

**Comment by JourneyOver** [comment]
@Cidan yep got it already as I had seen it was actually pushed right after I typed that above message xD everything looks to be back in working order now :D

**Pull request merged**